File: src/RSA.py
'''
Created on Mar 31, 2010

@author: oc
'''
from DynamInt import *
class RSA(object):
    '''
    classdocs
    '''


    def __init__(self, len):
        '''
        Constructor
        '''
        'number of bits for number'
        if len:
            self.len=len
        else:
            self.len=0; 
        'Initialization'
        self.p = DynamInt()
        self.q = DynamInt()
        self.d = DynamInt()
        self.e = DynamInt()
        self.phi=DynamInt()
        self.n=DynamInt()
        'Convenience numbers'
        self.zero=DynamInt()
        self.zero.setData("0")
        self.one=DynamInt()
        self.one.setData("1")
        self.two = DynamInt()
        self.two.setData("2")
        self.three = DynamInt()
        self.three.setData("3")
        self.base = DynamInt()
        self.base.setData("20")
    '''
    Calculating RSA Parameters4
    '''    
    def createkeys(self,p,q):
        
        while self.millerRabin(p) == False:
            'p or q is not prime, so choose another p or q'
            print "P is not prime  :  "+p.data
            p+=self.three
        while self.millerRabin(q) == False:
            print "Q is not prime  :  "+q.data
            q+=self.three
        
        #p = self.randomprime()
        #q = self.randomprime()   
         
        
        'numbers are prime so lets save the data and continue generating keys'
        self.p = p
        self.q = q
        self.n = p*q        
        self.createphi()
        print " Prime P        :   "+ p.data
        print " Prime Q        :   "+ q.data
        print " Modulus n      :   "+self.n.data
        e = DynamInt(1)
        '''
        Lets get a prime 1 < e < phi
        '''
        while not  (self.gcd(e,self.phi) == self.one):
            #e.random(len)
            e = e+self.one
            if e > self.phi: e = self.one
            #e.printme()
            #temp =self.gcd(e,self.phi)
        self.e = e
        '''
        constraint, 1 < d < phi
                    d = 1/e mod phi
        '''
        self.d = self.multiplicativeInverse(self.e, self.phi)
           
        print " encryption key :  "+ self.e.data
        print " decryption key :  "+ self.d.data
        return True
    def randomprime(self):
        '''
        randomprime
        generates a random prime by looping
        and testing each random number it creates
        '''
        print "Initial Prime"
        prime =  DynamInt(len)
        
        prime.printme()
        while self.isPrime(prime)!=True :
            print "Current Number "+prime.data          
            prime.random(len)
        return prime
    def getdecryptkey(self):
        self.d = self.extendedEuclid(self.phi, self.e)
        return False
    def messageToNum(self,m):
        '''
        converts message to numbers
        
        @param param m: message
        @return: dynamic integer with numbers representing message
        '''
        coded=self.disguise(m)
        return coded
    def numToMessage(self,m):
        '''
        converts numbers to message
        
        @param param m: message
        @return: dynamic integer with numbers representing message
        '''
        
        text=self.guise(m)
        
        return text
    def createphi(self):
        '''
        Create phi = (p-1)(q-1)
        '''
        self.phi=(self.p-self.one)*(self.q-self.one)
    def gcd(self,a,b):
        '''
        find the gcd between two numbers
        Uses the fact that 
        b = b%a*k + r next iteration b = r a = k, r = r%k*k2 + r2 
        example a = 10 b = 20
        on first loop
        while 10
            a = 20%10 = 0
            b = 10    
        hence returns 10    
        @param a, b for example e and phi
        '''
        t =  DynamInt()
        while not a == self.zero:
                t = a
                tr = b
                
                a, b = b%a, a
                tt = b/a
                #print tr.data +" = " + tt.data + "*"+t.data+"+"+a.data
        return b

    def isPrime(self, num):
        '''
        checks if a number is prime using the
        selected method
        for now method is millerRabin
        '''
        return self.millerRabin(num)
    def millerRabin(self, n):
        '''
        @function millerRabin
        encrypt
        by using modular exponention
        @param num  number to test for primality
        @return: True or False for primality
        '''
        isPrime = False

        if n%self.two == self.zero and n != self.two:
            return isPrime
        elif (n == self.zero or n == self.one or n == self.two or n == self.three):
            isPrime = True
            return isPrime
        else:
            p = n-self.one;
            s = self.zero;
            r = self.zero;
            
            while (p%self.two == self.zero):
            
                s+=self.zero
                p=p/self.two
            
            r = p
            #r = p/pow(2,s); //this may not work with BigInt but should
            i = self.one
            while i < s or i == s:
            
                #generate random a where 2<=a<=n-2 here
                #while >n-2 or <2 keep generating random numbers
                a = self.zero;
                while (a > (n-self.two) or a < self.two):
                    tt = DynamInt()
                    tt.random(n.size)
                    a = tt % n
                
                y = self.powmod(a, r, n)
                if not y == self.one and not y == n-self.one:
                
                    j=self.one;
                    while j < s-self.one or j == s-self.one and not y == n-self.one :
                    
                        y = (y*y)%n;
                        if (y == self.one):
                        
                            isPrime = False
                            return isPrime
                        
                        j+=self.one
                    
                    if (not y == n-self.one):
                    
                        isPrime = False;
                        return isPrime;
            isPrime = True;
            return isPrime;
    def powmod(self, a, r, n):
        return self.modularExp(a, r, n)
    def modularExp(self, a, r, n):
        '''
        function modularExp
        Calculates the modular exponentiation
        @param a,r,n
        '''
        'self.one is a dynamic integer of one, 1'
        result = self.one;

        while (r > self.zero):
        
            if not (r%self.two == self.zero):
            
                result = (result * a) % n;
            
    
            r = r/self.two;
            a = (a * a) % n;
        
        
        return result;
    def egcd(self,a, b):
        
        x, Xprev = self.zero, self.one
        y, Yprev = self.one, self.zero
     
        while (b>self.zero):
            quotient = a / b
            a, b = b, a % b
            x, Xprev = Xprev - quotient*x, x
            y, Yprev = Yprev - quotient*y, y
     
        return (Xprev, Yprev, a)
    def multiplicativeInverse(self,a, m):
        x, q, gcd = self.egcd(a, m)
     
        if gcd == self.one:
            return (x + m) % m
        else:
            return None
    '''Doing Encryption'''
    def encrypt(self, m):
        '''
        Encrypt the message using the keys created
        @param m: message
        '''
        c = self.powmod(self.messageToNum(m),self.e,self.n)
        return c
    def decrypt(self, m):
        m_ret = self.powmod(m,self.d,self.n)
        return self.reveal(m_ret)
    def disguise(self,text):
        '''
        converts message to numbers
        Accomplish this by looping through each element
        in message var and converting it int. 
        This int is added to a dynamint
        @param param m: message
        @return: dynamic integer with numbers representing message
        '''
        
        guise =self.zero 
        for i in text:
            'convert to ascii base 10'
            ascii = DynamInt()
            ascii.setData(str(ord(i)))
            guise = ascii+(guise*self.base)
        return guise
    def reveal(self,num):
        '''
        @summary: Converts from base 10 to readable string
        @param num: number to reveal by converting from base 10 to string (8 bit data) 
        '''
        g = []
        
        while num >self.zero:
            r , num = num % self.base, num / self.base
            temp = r-self.one
            g.append(string.lowercase[int(temp.data)])
        g.reverse()
        return g
    

File: src/rsa/RSA.cpp
#include "RSA.hpp"

/**
Created on Mar 31, 2010

@author: oc
*/

namespace __RSA__ {

str *const_0, *const_1, *const_2;

str *__name__;


/**
class RSA
*/

class_ *cl_RSA;

DynamInt *RSA::powmod(DynamInt *a, DynamInt *r, DynamInt *n) {
    
    return this->modularExp(a, r, n);
}

void *RSA::messagetonum(str *m) {
    /**
    converts message to numbers
    Accomplish this by looping through each element
    in message var and converting it int. 
    This int is added to a dynamint
    @param param m: message
    @return: dynamic integer with numbers representing message
    */
    __iter<str *> *__3;
    str *__2, *__ss_char;
    int __4;
    DynamInt *temp;

    temp = (new DynamInt(len(m)));

    FOR_IN(__ss_char,m,3)
        temp = temp->__add__(ord(__ss_char));
    END_FOR

    return NULL;
}

__ss_bool RSA::millerRabin(DynamInt *num) {
    /**
    @function millerRabin
    compute decryption key
    by using modular exponention
    @param num  number to test for primality
    @return: True or False for primality
    */
    DynamInt *n;

    n = 0;
    n = num;
    return True;
}

DynamInt *RSA::gcd(DynamInt *a, DynamInt *b) {
    /**
    find the gcd between two numbers
    Uses the fact that 
    b = b%a*k + r next iteration b = r a = k, r = r%k*k2 + r2 
    example a = 10 b = 20
    on first loop
    while 10
        a = 20%10 = 0
        b = 10    
    hence returns 10    
    @param a, b for example e and phi
    */
	DynamInt *__5;
    DynamInt *__6;


    while (__gt(a, this->one)) {
        __5 = b->__mod__(a);
        __6 = a;
        a = __5;
        b = __6;
    }
    return b;
}

DynamInt *RSA::decrypt(str *m) {
    DynamInt *m_ret;

    m_ret = (new DynamInt(((lambda0 )(0))));
    m_ret = ((DynamInt *)(this->powmod(((DynamInt *)(this->messagetonum(m))), this->d, ((DynamInt *)(this->n)))));
    return m_ret;
}

DynamInt *RSA::modularExp(DynamInt *a, DynamInt *r, DynamInt *n) {
    /**
    function modularExp
    Calculates the modular exponentiation
    @param a,r,n
    */
    
    return (new DynamInt(((lambda0 )(0))));
}

__ss_bool RSA::getdecryptkey() {
    
    this->d = this->extendedEuclid(this->phi, this->e);
    return False;
}

str *RSA::encrypt(str *m) {
    /**
    Encrypt the message using the keys created
    @param m: message
    */
    DynamInt *c;

    c = (new DynamInt(((lambda0 )(0))));
    c = ((DynamInt *)(this->powmod(((DynamInt *)(this->messagetonum(m))), this->e, ((DynamInt *)(this->n)))));
    return c;
}

__ss_bool RSA::createkeys(DynamInt *p, DynamInt *q) {
    /**
    check of we can p and q are prime
    */
    __ss_bool __0, __1;
    DynamInt *e;

    if (((this->millerRabin(p)==False) or (this->millerRabin(q)==False))) {
        /**
        p or q is not prime, so choose another p or q
        */
        print(1, const_0);
        return False;
    }
    /**
    numbers are prime so lets save the data and continue generating keys
    */
    this->p = p;
    this->q = q;
    this->n = p->__mul__(q);
    this->phi = ((DynamInt *)(this->createphi()));
    /**
    generate a random encrypt key of length len
    */
    e = (new DynamInt(((lambda0 )(__lambda0__))));

    while (__gt((new DynamInt(this->gcd(e, this->phi))), this->one)) {
        e->random(((lambda1 )(__lambda1__)));
    }
    this->e = e;
    return True;
}

void *RSA::createphi() {
    /**
    Create phi = (p-1)(q-1)
    */
    
    this->phi = ((DynamInt *)(((this->p)->__sub__(this->one))->__mul__((this->q)->__sub__(this->one))));
    return NULL;
}

void *RSA::__init__(int len) {
    /**
    Constructor
    */
    
    /**
    number of bits for number
    */
    if (len) {
        this->len = len;
    }
    else {
        this->len = 0;
    }
    /**
    Initialization
    */
    this->p = (new DynamInt(((lambda0 )(0))));
    this->q = (new DynamInt(((lambda0 )(0))));
    this->d = (new DynamInt(((lambda0 )(0))));
    this->e = (new DynamInt(((lambda0 )(0))));
    this->phi = (new DynamInt(((lambda0 )(0))));
    /**
    Convineince numbers
    */
    this->zero = (new DynamInt(((lambda0 )(0))));
    this->zero->data = const_1;
    this->one = (new DynamInt(((lambda0 )(0))));
    this->one->data = const_2;
    return NULL;
}

DynamInt *RSA::extendedEuclid(DynamInt *A, DynamInt *B) {
    /**
    @function extendedEuclid
    finding the greatest common divisor 
    @param A, B, x, y 
    */
    DynamInt *temp;

    temp = (new DynamInt(((lambda0 )(0))));
    return temp;
}

void __init() {
    const_0 = new str("P or Q is not prime!");
    const_1 = new str("0");
    const_2 = new str("1");

    __name__ = new str("RSA");

    cl_RSA = new class_("RSA", 33, 33);
}

} // module namespace

